### DISCUSSION TITLE
Overhaul the CHANGELOG.rst file.

### DISCUSSION THREAD

**Discussion by Elliria** [autokey/autokey#895] [Maintenance goals]
This is long overdue and is a monster task, so it needs to be broken down into pieces. Below are the necessary tasks with explanations and some of the hurdles that may arise. As always, I'm interested in feedback, suggestions, and corrections:

* Since this file is necessarily different on each of our three main branches, the changes will need to be made individually to the [beta](https://github.com/autokey/autokey/blob/beta/CHANGELOG.rst), [develop](https://github.com/autokey/autokey/blob/develop/CHANGELOG.rst), and [master](https://github.com/autokey/autokey/blob/master/CHANGELOG.rst) branches. This job will be made somewhat easier since all three files are identical from **Version 0.95.10 <2019-12-16>** down.
* Even though I'm aware (as [mentioned in this pull request](https://github.com/autokey/autokey/pull/892#issuecomment-1664281304)) that every effort needs to be made to preserve each entry in the `CHANGELOG.rst` file as is, some entries will need to be altered slightly, as described in that message and below, however it's a small number of the entries and won't prevent them from being used for searches, but will make the searches more challenging.
* Some of the entries take up multiple lines with blank lines between them. This is something that's difficult to achieve in lists with RST documents, so since the bulk of this file is in list format, the best approach would be to convert those to one-liners. This will probably involve changing some of the text, but once again, it's a small number of the entries and won't prevent them from being used for searches, but will make the searches more challenging.
* 🛑 The `python3 setup.py test` code snippet is missing its closing backtick, in [line 113 of the CHANGELOG.rst](https://github.com/autokey/autokey/blob/master/CHANGELOG.rst#L113) file on the **master** branch (this has been fixed on the **beta** and **develop** branches), which breaks the page, so this will be fixed. This will involve changing the text, however it's one entry and won't prevent it from being used for searches, but will make the search more challenging.
* RST documents are forgiving about the syntax used to create the headings being too long, which some of ours are, but since that's inconsistent, it will be cleaned up as part of this overhaul.
* 🛑 RST documents are unforgiving about the syntax used to create the headings being too short, which several of ours are and which breaks the file, so those will be cleaned up.
* 🛑 RST documents are unforgiving about how we travel through the heading levels and we've broken its rules, so our page is broken and will be repaired. The way headings work in RST is like this:
  * If the current heading is a level 3 heading:
    * If the level of the next heading below the current heading is the same number as the current one (if it's a level 3 heading), that's fine.
      * You can do this (dropping from a level 1 to a level 1 is fine):
        ```
        level 1
        level 1
        level 1
        ```
    * If the level of the next heading below the current heading is a lower number than the current one (if it's a level 1 or level 2 heading), that's fine. You can skip through these levels if you like.
      * You can do this (dropping from a level 3 to a level 1 is fine):
        ```
        level 1
          level 2
            level 3
        level 1
        ```
    * If the level of the next heading below the current heading is a higher level than the current one (if it's a level 4 or more), it **must** be the **very next level** (it must be a level 4). You cannot skip through these levels or the page will break.
      * You can do this (dropping from a level 3 to a level 4 is fine):
        ```
        level 1
          level 2
            level 3
              level 4
        ```
      * You **cannot** do this (dropping from a level 3 to a level 5 is forbidden):
        ```
        level 1
          level 2
            level 3
                level 5
         ```
* We have way too many types of headings, making the document busy and difficult to navigate. I've attached the [Headings.txt](https://github.com/autokey/autokey/files/12255053/Headings.txt) file with all of the headings in the [CHANGELOG.rst](https://github.com/autokey/autokey/blob/master/CHANGELOG.rst) file on the **master** branch in order of appearance. *Note that the different styles of underlining beneath each one represents its heading level, and if you're curious, note, also, the spots in which we jump to the wrong levels.* Those will be whittled down to a smaller, more manageable list. Here are the headings I propose that we use:
  * Breaking changes
  * Documentation
  * Features
  * Fixes
  * Other
  * Packaging
* Our headings have not been used consistently. In other words, sometimes bug fixes are listed first and sometimes other types of changes are listed first. Those will all be made consistent and will adhere to these rules:
  * Each section should only use necessary headings and shouldn't have any headings with no content beneath them.
  * The headings in each section should appear in alphabetical order for ease in searching.
* Only some of the links to issues or pull requests are clickable. The missing clickable links will be added. That will change the text, but only the reference to which issue it fixes or addresses and not the text that refers to the actual change that was made, which is what would be used in searches through the git history.

**Comment by josephj11** [comment]
This was a substantial amount of work already!

If you feel that preserving things to make searching easier is important, that makes sense. It would be clearer to me with an example search so  I can see what you're talking about. I haven't done much searching on GitHub. I do it on my local cloned copy with grep. It would be good to know how to do it without needing a local copy.

I think more in terms of major to minor rather than alphabetically ordered sections, but it's no big deal as long as whatever order we select is applied consistently.

What are "upgrades"? newer versions of Python and actions?

How did you get those red octagons to show up?

**Comment by Elliria** [comment]
It was even more work than it seems, because I've already done all of it, but deleted it all and started over here.

See the **Search** section of my [AutoKey useful links](https://gist.github.com/Elliria/cad19c1837e1b781edfcdb6d981987be) gist for lots of ways to search AutoKey on GitHub. The **commits** search is what you'd be after, although when you're doing a deep investigation, you'll want to search the code, the commits, the pull requests, etc.

I can understand why you might like the major or minor sorting, but it's harder to quickly search and is completely disorienting to those of us who expect alphabetical sorts, so I'd vote for alphabetical listing. It's how listings are done in pretty much every field everywhere, so it's what our brains tend to expect to encounter.

*As an example of why something like that is problematic, look at [my public gists](https://gist.github.com/Elliria). I don't have a lot of them compared to some folks, but there are enough of them that it's annoying to try to find the one you want because they're not sorted alphabetically. That's not offered as an option [sigh], so I'm stuck with them the way they are. I can sort them by when they were updated, so I could edit all of them in a specific order in order to force an alphabetic sort, but I'd have to repeat that every time I added a gist, which would be frustrating and shouldn't be necessary. I've either created or joined in on an issue about this (can't remember which), but nothing has been done yet. Here's hoping...*

**Upgrades** are badly named. I'm not sure how to refer to them. Maybe you can help me come up with a better name. This refers to the times when we make big leaps that make a change significant in comparison to most other changes. As an example, when we went from using Python 2 to using Python 3.which is a dramatic upgrade to the code-base, that would be considered an upgrade. If we were, for example, to switch from using **X** to using **Wayland**, that would also be considered an upgrade. Come to think of it, that category is unnecessary. After all, those two examples would fit into the **Breaking changes** category, which is for any change that will break things unless we manually intervene (by updating our scripts, for example). I'll get rid of **Upgrades** above.

I used the **Emoji selector** program that's in our **Utilities** menu in Kubuntu and copied the stop sign and pasted it into the message. We can use those all over GitHub. Here's another as an example: ❤️

**Reply by josephj11** [reply]
I will read your gist.

Simple solution: Create one (or more) gists that are indexes of the rest of them with URL links. You have to maintain it manually, but it's pretty easy especially with things like my md link from clipboard and selection script. I'm not sure where it works, but Ctrl+k creates a link from a selection as the label and leaves the cursor where the URL can be pasted.

A new feature can be orthogonal too. If it just does something new, then it doesn't break anything old. If you never use the `<code>` macro it doesn't matter (to you) that it was added.

These would probably come under new features or breaking changes depending on what they do. Some could be both and breaking changes would be the more important place to put those.

**Comment by Elliria** [comment]
The new Gist used as an index is a good idea, so [I've done it](https://gist.github.com/Elliria/3b2367ec18d4c927cd3c066464a5dc76). Thanks for the suggestion. It also decluttered my GitHub home page. I'm still hoping they'll fix the sort, though, because when it comes to editing the Gists, there will still be hunting necessary to find the one(s) you want to edit.

As far as orthogonal entries, those would be problematic, because each entry should only appear once. I'm hoping that that can be dealt with by significantly whittling down the number of possible categories. Also, sometimes a choice will just have to be made between one category and another. If it's the wrong one, entries are easy enough to move later.

I decided to fix the [blank issue](https://github.com/autokey/autokey/issues/897) issue first and will start on all of the changes mentioned above after that. I came up with two possible plans for doing that and can do either one, but have a feeling you'd prefer the alternative one. Each plan consists of a number of steps. Each step will become a separate issue in our tracker so that it can be handled independently and be referenced and linked to in the **CHANGELOG.rst** file:

### The original plan:
1. Fix the `python3 setup.py test` code snippet in [line 113 of the CHANGELOG.rst](https://github.com/autokey/autokey/blob/master/CHANGELOG.rst#L113) file on the **master** branch.
2. Make all links clickable.
3. Convert all entries to one-liners.
4. Choose new headings.
5. Adjust length of heading syntax.
6. Adjust heading levels, replace the headings with the new ones, and organize the entries under the new headings.

**Step 5** may not be necessary if the broken headings will no longer be used based on the choices made in **step 4**. The problem is going to be that **step 6** is what should be three separate steps done all at once in one commit (that cannot be done in separate commits because of the way the document was written and the unforgiving nature of RST files), which will be a dramatic change in the document that will be challenging to review when checking my work.

### A possible alternative plan:
The only way I can see around that would be to get rid of most of the headings (leaving us with just version headings and the entries beneath them) in one commit, and then creating new headings and using them by organizing the entries beneath them in the next commit. That would look like this:
1. Fix the `python3 setup.py test` code snippet in [line 113 of the CHANGELOG.rst](https://github.com/autokey/autokey/blob/master/CHANGELOG.rst#L113) file on the **master** branch.
2. Make all links clickable
3. Convert all entries to one-liners
4. Choose new headings.
5. Remove all headings except version headings.
6. Add new headings and organize the entries under them.

**Reply by josephj11** [reply]
I can't tell the effective differences between the two approaches.

This is what the bottom line is for me:

We have two types of changes: Changes to content/semantics and Changes to organization/structure.

It would be easier, whenever possible, if the first type of changes were done first. Then, most of the changes could be reviewed side by side (like diff does). This allows looking at them one at a time in context.

Once that's done, we know what we're saying is correct/agreed upon and can focus on any needed/desired reorganization.

**Comment by Elliria** [comment]
Only the first four steps in either scenario will give you that, and those are the ones I intend to do first, in the order provided. You should have an easy time reviewing the changes in those four steps.

Step 5 is different in both scenarios:
* It's optional in the first scenario and will depend on which headings were chosen in step 4. If the ones that are broken are on that list, they'll need to be adjusted in step 5 of that scenario.
* It's required in the second scenario, but will be easy for you to review.

Where we run into trouble is in step 6:
* In the **first** scenario, this step involves removals, changes, additions, and movement, making it messy and difficult to review.
* In the **second** scenario, this step involves additions and movement, which should be easier to review.

Step 6 will be rough in either case, because there's a lot of change happening, but more is happening in the first scenario. What's most important about all of this, though, is that we're not ever losing any of the entries. The only data we'll lose or gain are headings.

**Reply by josephj11** [reply]
Sounds like it's a choice between more work for you and more work for me. (I'm a bit biased  on that. LOL)

OTOH, this is primarily a one off occurrence, so it's not that important to solve it generically. Either way is fine. The first way is probably better because it's less work for you and only headings/structure are at risk.

**Comment by Elliria** [comment]
Thank you for thinking of me, but neither is less work for me. The file is a mess. Also, this discussion was opened for you because the pull request I did in #892 was so difficult to review that you gave up almost immediately. It makes sense to do it in steps, with each step being a commit, to make it gentler to follow.

Since steps 1 through 4 are identical in both scenarios, I'll do those on the **beta**, **develop**, and **master** branches individually (lots of commits). The file (in each branch) won't look much different than it does now, but we'll both be looking at exactly what it's giving us to work with.

Then, we can take it from there, and I'm basically going to be hitting you with either a sledgehammer or a mallet in the last step (possibly the last two steps, but we'll know that once we complete step 4). In either case, the diff will be misleading and will show additions and deletions rather than moves, so there's not going to be any way to get around going through both files line by line to find each one manually if you want to be certain that nothing got deleted.

How does that sound?

**Reply by josephj11** [reply]
That sounds as good as it can be. Maybe there will be an AI version soon that can recognize moves or even do it for us.

**Comment by Elliria** [comment]
Oh, I sincerely hope not. I'm going to be the last woman standing who's against the use of AI and is watching its negative impacts already beginning to unfold with great interest as it follows its inevitable course to gradually fulfill her predictions.

Also, depending on what's decided in issue #493, this may become a gentler process, because Markdown files would be more forgiving of heading levels and easier to compare in diffs.

**Reply by josephj11** [reply]
But its purpose is "[To serve man](https://en.wikipedia.org/wiki/To_Serve_Man_(The_Twilight_Zone))"!

**Reply by Elliria** [reply]
That's exactly the problem. There's nothing inherently wrong with the AI software (other than any expected or unexpected bugs that could happen to any software). It's a serviceable tool. Man, however, cannot be trusted to play well with tools.
